# **MSc Data Science thesis project: Micromobility vehicle parking**
## **Data pre-processing**

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

# Set main directories
main_path = 'C:/Users/jfieb/Documents/Data science/Thesis project/Data'
bikecount_path = main_path + '/bikecount_raw/'
shapefiles_path = main_path + '/shapefiles/'

In [2]:
# Read the (raw) vehicle parking count data from 2018, 2019, 2020 and 2021
raw_2018 = pd.read_csv(bikecount_path + '2018_raw.csv',delimiter=';',encoding = 'utf-8-sig')
raw_2019 = pd.read_csv(bikecount_path + '2019_raw.csv',delimiter=';',encoding = 'ISO-8859-1')
raw_2020 = pd.read_csv(bikecount_path + '2020_raw.csv',delimiter=';',encoding = 'utf-8-sig')
raw_2021 = pd.read_csv(bikecount_path + '2021_raw.csv',delimiter=';',encoding = 'utf-8-sig')

In [3]:
raw_2018.head(5)

id day_part  section_2018 letter   merged  section_2016  id_total  \
0  1199945   middag         35617     bv  35617bv       35617.0       NaN   
1  1199946   middag         35617      r   35617r       35617.0       NaN   
2  1199947   middag         35617      n   35617n       35617.0       NaN   
3  1199948   middag         35617      v   35617v       35617.0       NaN   
4  1199949   middag         35617     vn  35617vn       35617.0       NaN   

          sort  group  year  date  day  time_code     time_period  \
0       sectie    NaN  2018   NaN  NaN        NaN  13:00 - 14:00h   
1  voorziening    NaN  2018   NaN  NaN        NaN  13:00 - 14:00h   
2  voorziening    NaN  2018   NaN  NaN        NaN  13:00 - 14:00h   
3  voorziening    NaN  2018   NaN  NaN        NaN  13:00 - 14:00h   
4  voorziening    NaN  2018   NaN  NaN        NaN  13:00 - 14:00h   

               district ois_2015 ois_2018        street  category  district_2  \
0  Zuid (Buitenveldert)     K90g     K23c  De Boelelaan       NaN         NaN   
1  Zuid (Buitenveldert)     K90g     K23c  De Boelelaan       NaN         NaN   
2  Zuid (Buitenveldert)     K90g     K23c  De Boelelaan       NaN         NaN   
3  Zuid (Buitenveldert)     K90g     K23c  De Boelelaan       NaN         NaN   
4  Zuid (Buitenveldert)     K90g     K23c  De Boelelaan       NaN         NaN   

   facility_id            location            facility  cap_bike  cap_moped  \
0            0  buiten voorziening  Buiten voorziening         0        NaN   
1            1      in voorziening               Rek           0        NaN   
2            2      in voorziening           Nietjes          12        NaN   
3            3      in voorziening               Vak           0        NaN   
4            4      in voorziening     Vak met nietjes         0        NaN   

   cap_brokenup  cap_private cap_comment  bike  moped_yellow  moped_blue  \
0             0            0         NaN     0             0           0   
1             0            0         NaN     0             0           0   
2             0            0         NaN     7             1           1   
3             0            0         NaN     0             0           0   
4             0            0         NaN     0             0           0   

   cargo_bike motor  fourwheeler  stock  total_unmotorized  total_twowheeler  \
0           0     0            0      0                  0                 0   
1           0     0            0      0                  0                 0   
2           0     0            0      0                  7                 9   
3           0     0            0      0                  0                 0   
4           0     0            0      0                  0                 0   

    area measured_by  
0  321.6      Trajan  
1  321.6      Trajan  
2  321.6      Trajan  
3  321.6      Trajan  
4  321.6      Trajan

In [4]:
raw_2019.head(5)

id day_part  section_2018 letter   merged section_2016  id_total  \
0  1011881.0   middag         11494     bv  11494bv     109327.0       NaN   
1  1011882.0   middag         11494      r   11494r     109327.0       NaN   
2  1011883.0   middag         11494      n   11494n     109327.0       NaN   
3  1011884.0   middag         11494      v   11494v     109327.0       NaN   
4  1011885.0   middag         11494     vn  11494vn     109327.0       NaN   

          sort group  year  date  day  time_code     time_period district  \
0       sectie   NaN  2019   NaN  NaN        NaN  13:00 - 14:00h  Centrum   
1  voorziening   NaN  2019   NaN  NaN        NaN  13:00 - 14:00h  Centrum   
2  voorziening   NaN  2019   NaN  NaN        NaN  13:00 - 14:00h  Centrum   
3  voorziening   NaN  2019   NaN  NaN        NaN  13:00 - 14:00h  Centrum   
4  voorziening   NaN  2019   NaN  NaN        NaN  13:00 - 14:00h  Centrum   

  ois_2015 ois_2018             street  category district_2 district_nr  \
0     A09e     A09e  Czaar Peterstraat       NaN    centrum  centrum-78   
1     A09e     A09e  Czaar Peterstraat       NaN    centrum  centrum-78   
2     A09e     A09e  Czaar Peterstraat       NaN    centrum  centrum-78   
3     A09e     A09e  Czaar Peterstraat       NaN    centrum  centrum-78   
4     A09e     A09e  Czaar Peterstraat       NaN    centrum  centrum-78   

   facility_id            location            facility  cap_bike  cap_moped  \
0            0  Buiten voorziening  Buiten voorziening       0.0        NaN   
1            1      in voorziening               Rek         0.0        NaN   
2            2      in voorziening           Nietjes        48.0        NaN   
3            3      in voorziening               Vak         0.0        NaN   
4            4      in voorziening         Gemengd vak       0.0        NaN   

   cap_brokenup  cap_private cap_comment  bike  moped_yellow  moped_blue  \
0           0.0          0.0         NaN   1.0           0.0         0.0   
1           0.0          0.0         NaN   0.0           0.0         0.0   
2           0.0          0.0         NaN  35.0           0.0         0.0   
3           0.0          0.0         NaN   0.0           0.0         0.0   
4           0.0          0.0         NaN   0.0           0.0         0.0   

   cargo_bike  motor  fourwheeler  stock  total_unmotorized  total_twowheeler  \
0         0.0    0.0          0.0    0.0                  1                 1   
1         0.0    0.0          0.0    0.0                  0                 0   
2         1.0    0.0          0.0    0.0                 36                36   
3         0.0    0.0          0.0    0.0                  0                 0   
4         0.0    0.0          0.0    0.0                  0                 0   

    area measured_by  
0  351.5      Trajan  
1  351.5      Trajan  
2  351.5      Trajan  
3  351.5      Trajan  
4  351.5      Trajan

In [5]:
raw_2020.head(5)

section_2018        street              neighborhood   ois  year day_part  \
0        100000  Weesperzijde  Weesperzijde Midden/Zuid  M27b  2020    avond   
1        100000  Weesperzijde  Weesperzijde Midden/Zuid  M27b  2020    avond   
2        100000  Weesperzijde  Weesperzijde Midden/Zuid  M27b  2020    avond   
3        100000  Weesperzijde  Weesperzijde Midden/Zuid  M27b  2020    avond   
4        100000  Weesperzijde  Weesperzijde Midden/Zuid  M27b  2020    avond   

  horeca_ov               facility  cap_bike  cap_moped  cap_brokenup  \
0       NaN  0. Buiten voorziening         0          0             0   
1       NaN                 1. Rek         0          0             0   
2       NaN             2. Nietjes         2          0             0   
3       NaN                 3. Vak         0          0             0   
4       NaN         4. Gemengd vak         0          0             0   

   cap_private cap_comment  bike  moped_yellow  moped_blue  cargo_bike  motor  \
0            0         NaN     6             0           0           0      0   
1            0         NaN     0             0           0           0      0   
2            0         NaN     2             0           0           0      0   
3            0         NaN     0             0           0           0      0   
4            0         NaN     0             0           0           0      0   

   fourwheeler  stock  rent_shared_bike  brokenup  sum_cap  sum_count  
0            0      0                 0         0        0          6  
1            0      0                 0         0        0          0  
2            0      0                 0         0        2          2  
3            0      0                 0         0        0          0  
4            0      0                 0         0        0          0

In [6]:
raw_2021.head(5)

section_2018                    street      neighbourhood   ois district  \
0         10001           Ooster Ringdijk  Science Park Zuid  M56i     Oost   
1         10002  Westel Merwedekanaaldijk  Science Park Zuid  M56i     Oost   
2         10003           Ooster Ringdijk  Science Park Zuid  M56i     Oost   
3         10004  Westel Merwedekanaaldijk  Science Park Zuid  M56i     Oost   
4         10005           Ooster Ringdijk  Science Park Zuid  M56i     Oost   

   year day_part    time_period               facility  cap_bike  cap_moped  \
0  2021   middag  12:00 - 16:00  0. Buiten voorziening         0          0   
1  2021   middag  12:00 - 16:00  0. Buiten voorziening         0          0   
2  2021   middag  12:00 - 16:00  0. Buiten voorziening         0          0   
3  2021   middag  12:00 - 16:00  0. Buiten voorziening         0          0   
4  2021   middag  12:00 - 16:00  0. Buiten voorziening         0          0   

   cap_brokenup  cap_private  bike  moped_yellow  moped_blue  cargo_bike  \
0             0            0     0             0           0           0   
1             0            0     1             0           0           0   
2             0            0     0             0           0           0   
3             0            0     0             0           0           0   
4             0            0     0             0           0           0   

   motor  fourwheeler  stock  rent_shared_bike  brokenup  sum_cap  sum_count  
0      0            0      0                 0         0        0          0  
1      0            0      0                 0         0        0          1  
2      0            0      0                 0         0        0          0  
3      0            0      0                 0         0        0          0  
4      0            0      0                 0         0        0          0

In [7]:
print('Shapes')
print('2018: ' + str(raw_2018.shape))
print('2019: ' + str(raw_2019.shape))
print('2020: ' + str(raw_2020.shape))
print('2021: ' + str(raw_2021.shape))

Shapes
2018: (260432, 39)
2019: (124767, 40)
2020: (103208, 24)
2021: (23996, 24)


### **Shapes**
The datasets contain different numbers of rows and columns:
- 2018: 260432 rows and 29 columns
- 2019: 124767 rows and 40 columns
- 2020: 103208 rows and 24 columns
- 2021: 23996 rows and 24 columns

In [8]:
print('Features:')
print('2018')
print(str(raw_2018.columns))
print('2019')
print(str(raw_2019.columns))
print('2020')
print(str(raw_2020.columns))
print('2021')
print(str(raw_2021.columns))

Features:
2018
Index(['id', 'day_part', 'section_2018', 'letter', 'merged', 'section_2016',
       'id_total', 'sort', 'group', 'year', 'date', 'day', 'time_code',
       'time_period', 'district', 'ois_2015', 'ois_2018', 'street', 'category',
       'district_2', 'facility_id', 'location', 'facility', 'cap_bike',
       'cap_moped', 'cap_brokenup', 'cap_private', 'cap_comment', 'bike',
       'moped_yellow', 'moped_blue', 'cargo_bike', 'motor', 'fourwheeler',
       'stock', 'total_unmotorized', 'total_twowheeler', 'area',
       'measured_by'],
      dtype='object')
2019
Index(['id', 'day_part', 'section_2018', 'letter', 'merged', 'section_2016',
       'id_total', 'sort', 'group', 'year', 'date', 'day', 'time_code',
       'time_period', 'district', 'ois_2015', 'ois_2018', 'street', 'category',
       'district_2', 'district_nr', 'facility_id', 'location', 'facility',
       'cap_bike', 'cap_moped', 'cap_brokenup', 'cap_private', 'cap_comment',
       'bike', 'moped_yellow', 'mop

### **Features**
The datasets contain different features. The most important/interesting features are the following:
- section_2018: The identification of the observed sidewalk section
- year: The year the observation was made
- day_part: The part of the day when the observation was made
- facility: The type of the observed parking facility (e.g. rack, staples, space)
- cap_bike: The bike parking capacity of the observed facility
- cap_moped: The moped parking capacity of the observed facility
- cap_brokenup: The parking capacity of the observed facility for a sidewalk segment with roadworks
- cap_private: The private parking capacity of the observed facility (e.g. a wall ring) 
- bike: The bicycle parking occupancy of the observed facility
- moped_yellow: The moped (yellow) parking occupancy of the observed facility
- moped_blue: The moped (blue) parking occupancy of the observed facility
- cargo_bike: The cargo bicycle parking occupancy of the observed facility
- motor: The motor parking occupancy of the observed facility
- four_wheeler: The four wheeler parking occupancy of the observed facility
- stock: The stock bicycle parking occupancy of the observed facility
- brokenup: The bicycle parking occupancy of the observed facility for a sidewalk segment with roadworks
- rent_shared_bike: The parking occupancy of shared bicycles for the observed facility

The 2018 and 2019 datasets have no features for the parking occupancy of shared vehicles (rent_shared_bike) and vehicles at locations with roadworks (brokenup). It is unclear whether the parking counts for those vehicles are included in other vehicle parking features or if they simply were not counted. 

To make the processing of the data easier, it was decided to add the missing attributes to the 2018 and 2019 data, with the parking counts set to 0.

In [9]:
# Add missing columns to the data from 2018 and 2019
raw_2018['rent_shared_bike'] = 0
raw_2019['rent_shared_bike'] = 0
raw_2018['brokenup'] = 0
raw_2019['brokenup'] = 0

### **Create subsets**

Subsets of the data are created using the key features, which indicate the location and time of the observation and the parking capacity and occupancy for different types of vehicles and different types of parking facilities.

In [10]:
# Define a subset of features
columns = ['section_2018','year','day_part', 'facility', 'cap_bike','cap_moped', 'cap_brokenup', 'cap_private', 'bike',
       'moped_yellow', 'moped_blue', 'cargo_bike', 'motor', 'fourwheeler','rent_shared_bike', 'stock', 'brokenup']

# Create subsets of the data
data_2018 = raw_2018[columns]
data_2019 = raw_2019[columns]
data_2020 = raw_2020[columns]
data_2021 = raw_2021[columns]

In [11]:
print('Shapes of data subsets')
print('2018: ' + str(data_2018.shape))
print('2019: ' + str(data_2019.shape))
print('2020: ' + str(data_2020.shape))
print('2021: ' + str(data_2021.shape))

Shapes of data subsets
2018: (260432, 17)
2019: (124767, 17)
2020: (103208, 17)
2021: (23996, 17)


Now all datasets contain the same number of 17 features. 

### **Facility feature**

The values for the feature 'facility' differ between the 2018/2019 and 2020/2021 datasets. New features are added to each dataset to align the values.

In [12]:
print('Unique values for facility feature')
print('2018')
print(str(data_2018['facility'].unique()))
print('2019')
print(str(data_2019['facility'].unique()))
print('2020')
print(str(data_2020['facility'].unique()))
print('2021')
print(str(data_2021['facility'].unique()))

Unique values for facility feature
2018
['Buiten voorziening' 'Rek  ' 'Nietjes  ' 'Vak  ' 'Vak met nietjes'
 'Bromfietsvak  ' 'Voor fietsenwinkel ' 'Overig  ' nan]
2019
['Buiten voorziening' 'Rek  ' 'Nietjes  ' 'Vak  ' 'Gemengd vak'
 'Bromfietsvak  ' 'Voor fietsenwinkel ' 'Overig  ' 'Vak met nietjes']
2020
['0. Buiten voorziening' '1. Rek' '2. Nietjes' '3. Vak' '4. Gemengd vak'
 '5. Bromfietsvak' '6. Voor fietsenwinkel' '7. Overig']
2021
['0. Buiten voorziening' '1. Rek' '2. Nietjes' '5. Bromfietsvak'
 '7. Overig' '3. Vak' '4. Gemengd vak' '6. Voor fietsenwinkel']


In [13]:
# Use substrings of the facility values for the 2020 and 2021 datasets to remove the numbers
data_2020['facility'] = data_2020['facility'].str[3:]
data_2021['facility'] = data_2021['facility'].str[3:]

In [14]:
# Add features that describe the facility type and location
def add_facility(df):
    df['facility'] = df['facility'].str.rstrip()

    conditions = [
    df['facility'] == "Buiten voorziening",
    df['facility'] == "Rek",
    df['facility'] == "Nietjes",
    df['facility'] == "Vak",
    df['facility'] == "Bromfietsvak",
    df['facility'] == "Voor fietsenwinkel",
    df['facility'] == "Gemengd vak",
    df['facility'] == "Overig",
    df['facility'] == "Vak met nietjes"]
    choices_type = ['no_facility','rack', 'staples','space','moped_space','bike_shop', 'mixed_space','other','space_with_staples']
    choices_location = ['outside','inside', 'inside','inside','inside','inside', 'inside','inside','inside']
    df['facility_type'] = np.select(conditions, choices_type, default = None)
    df['facility_location'] = np.select(conditions, choices_location, default = None)

    return df

In [15]:
data_2018 = add_facility(data_2018)
data_2019 = add_facility(data_2019)
data_2020 = add_facility(data_2020)
data_2021 = add_facility(data_2021)

In [16]:
print('2018: ' + str(data_2018.shape))
print('2019: ' + str(data_2019.shape))
print('2020: ' + str(data_2020.shape))
print('2021: ' + str(data_2021.shape))

2018: (260432, 19)
2019: (124767, 19)
2020: (103208, 19)
2021: (23996, 19)


### **Check for missing values**

In [17]:
# Definition to check for missing values in the datasets
def check_for_missing():
    null_2018 = pd.Series(data_2018.isnull().sum(), name = '2018')
    null_2019 = pd.Series(data_2019.isnull().sum(), name = '2019')
    null_2020 = pd.Series(data_2020.isnull().sum(), name = '2020')
    null_2021 = pd.Series(data_2021.isnull().sum(), name = '2021')
    missing = pd.concat([null_2018, null_2019, null_2020, null_2021], axis=1)
    return missing

In [18]:
check_for_missing()

2018    2019  2020  2021
section_2018            0       0     0     0
year                    0       0     0     0
day_part                0       0     0     0
facility             4760       0     0     0
cap_bike                0       4     0     0
cap_moped          260432  123015     0     0
cap_brokenup            0     128     0     0
cap_private             0     128     0     0
bike                    0     623     0     0
moped_yellow            0     623     0     0
moped_blue              0     623     0     0
cargo_bike              0     623     0     0
motor                   0     623     0     0
fourwheeler             0     623     0     0
rent_shared_bike        0       0     0     0
stock                   0     623     0     0
brokenup                0       0     0     0
facility_type        4760       0     0     0
facility_location    4760       0     0     0

**Missing values**

2018
- There are 4760 missing values in the column ‘facility’,'facility_type' and 'facility_location' -> After checking the data it turns out that the observations have the location 'outside facility'. Therefore the missing values are filled according to this.
- There are 260432 missing values in the column ‘cap_moped’ (no values at all for this feature) → the question is whether there was actually no capacity for mopeds or whether there was capacity but it has not been counted ???

2019
- There are 4 missing values in column 'cap_bike', 123015 missing values in column ‘cap_moped’ (nearly no values at all for this feature) and 128 missing values for columns 'cap_brokenup' and 'cap_private'
- There are 623 observations with no parking occupancy counts at all ( = missing values the features: bike, moped_yellow, moped_blue, cargo_bike, motor_fourwheeler, stock) → the question is whether there have been any observations here at all or whether there have been no parked vehicles ??? 
   - Parking capacity counts are available, so that may indicate that the locations were actually counted, but there were no vehicles

2020 and 2021
- There are no missing values in the 2020 and 2021 datasets


In [19]:
# Fill missing facility values in 2018 dataset
data_2018['facility'].fillna('Buiten voorziening', inplace=True)
data_2018['facility_type'].fillna('no_facility', inplace=True)
data_2018['facility_location'].fillna('outside', inplace=True)

In [20]:
# Drop the 623 observations with missing parking counts in 2019 dataset
data_2019.dropna(subset = ['bike'], inplace = True)

In [21]:
# Create df with mean cap_moped of 2020 and 2021 data to fillna in 2018 and 2019 data
mean_cap_moped = pd.merge(data_2020[['section_2018','facility_type','cap_moped']], data_2021[['section_2018','facility_type','cap_moped']], how = 'outer', on= ['section_2018','facility_type'])
mean_cap_moped['cap_moped_mean'] = mean_cap_moped[['cap_moped_x','cap_moped_y']].mean(axis=1)
mean_cap_moped

section_2018 facility_type  cap_moped_x  cap_moped_y  cap_moped_mean
0             100000   no_facility          0.0          0.0             0.0
1             100000          rack          0.0          NaN             0.0
2             100000       staples          0.0          NaN             0.0
3             100000         space          0.0          NaN             0.0
4             100000   mixed_space          0.0          NaN             0.0
...              ...           ...          ...          ...             ...
106833        490236          rack          NaN          0.0             0.0
106834        490236   moped_space          NaN          2.0             2.0
106835        490257   no_facility          NaN          0.0             0.0
106836        490263   no_facility          NaN          0.0             0.0
106837        490280   no_facility          NaN          0.0             0.0

[106838 rows x 5 columns]

In [22]:
# Fill missing values for cap_moped in the 2018 and 2019 dataset with the mean of the 2020/2021 data
data_2018 = pd.merge(data_2018,mean_cap_moped[['section_2018','facility_type','cap_moped_mean']], how = 'left', on= ['section_2018','facility_type'])
data_2019 = pd.merge(data_2019,mean_cap_moped[['section_2018','facility_type','cap_moped_mean']], how = 'left', on= ['section_2018','facility_type'])
data_2018['cap_moped'].fillna(data_2018['cap_moped_mean'], inplace=True)
data_2019['cap_moped'].fillna(data_2019['cap_moped_mean'], inplace=True)
data_2018.drop(['cap_moped_mean'], axis=1, inplace=True)
data_2019.drop(['cap_moped_mean'], axis=1, inplace=True)
# Fill the rest of the missing values with 0
data_2018.fillna(0, inplace=True)
data_2019.fillna(0, inplace=True)

In [23]:
check_for_missing()

2018  2019  2020  2021
section_2018          0     0     0     0
year                  0     0     0     0
day_part              0     0     0     0
facility              0     0     0     0
cap_bike              0     0     0     0
cap_moped             0     0     0     0
cap_brokenup          0     0     0     0
cap_private           0     0     0     0
bike                  0     0     0     0
moped_yellow          0     0     0     0
moped_blue            0     0     0     0
cargo_bike            0     0     0     0
motor                 0     0     0     0
fourwheeler           0     0     0     0
rent_shared_bike      0     0     0     0
stock                 0     0     0     0
brokenup              0     0     0     0
facility_type         0     0     0     0
facility_location     0     0     0     0

There are no missing values left in the datasets

### **Check the data types**

In [24]:
data_2018.dtypes

section_2018           int64
year                   int64
day_part              object
facility              object
cap_bike               int64
cap_moped            float64
cap_brokenup           int64
cap_private            int64
bike                   int64
moped_yellow           int64
moped_blue             int64
cargo_bike             int64
motor                 object
fourwheeler            int64
rent_shared_bike       int64
stock                  int64
brokenup               int64
facility_type         object
facility_location     object
dtype: object

In [25]:
# Fix 'motor' feature in the 2018 dataset
data_2018['motor'] = data_2018['motor'].replace(' ', 0)
data_2018['motor'] = pd.to_numeric(data_2018['motor'])
# Set datatype to int
data_2018 = data_2018.astype({'cap_moped': 'int64'})

In [26]:
data_2019.dtypes

section_2018           int64
year                   int64
day_part              object
facility              object
cap_bike             float64
cap_moped            float64
cap_brokenup         float64
cap_private          float64
bike                 float64
moped_yellow         float64
moped_blue           float64
cargo_bike           float64
motor                float64
fourwheeler          float64
rent_shared_bike       int64
stock                float64
brokenup               int64
facility_type         object
facility_location     object
dtype: object

In [27]:
# Set all datatypes to int
data_2019 = data_2019.astype({'cap_bike': 'int64', 'cap_moped': 'int64', 'cap_brokenup': 'int64',
                             'cap_private': 'int64', 'bike': 'int64', 'moped_yellow': 'int64',
                             'moped_blue': 'int64', 'cargo_bike': 'int64', 'motor': 'int64',
                             'fourwheeler': 'int64', 'stock': 'int64'})

In [28]:
data_2020.dtypes

section_2018          int64
year                  int64
day_part             object
facility             object
cap_bike              int64
cap_moped             int64
cap_brokenup          int64
cap_private           int64
bike                  int64
moped_yellow          int64
moped_blue            int64
cargo_bike            int64
motor                 int64
fourwheeler           int64
rent_shared_bike      int64
stock                 int64
brokenup              int64
facility_type        object
facility_location    object
dtype: object

In [29]:
data_2021.dtypes

section_2018          int64
year                  int64
day_part             object
facility             object
cap_bike              int64
cap_moped             int64
cap_brokenup          int64
cap_private           int64
bike                  int64
moped_yellow          int64
moped_blue            int64
cargo_bike            int64
motor                 int64
fourwheeler           int64
rent_shared_bike      int64
stock                 int64
brokenup              int64
facility_type        object
facility_location    object
dtype: object

### **Add features**

In [30]:
# Add features with total capacity and occupancy to each dataset
def add_features(df):
    df['capacity'] = df[['cap_bike', 'cap_moped','cap_brokenup', 'cap_private']].sum(axis=1)
    df['occupancy'] = df[['bike', 'moped_yellow', 'moped_blue','cargo_bike', 'motor', 'fourwheeler', 'rent_shared_bike', 'stock', 'brokenup']].sum(axis=1)
    return df

In [31]:
data_2018 = add_features(data_2018)
data_2019 = add_features(data_2019)
data_2020 = add_features(data_2020)
data_2021 = add_features(data_2021)

In [32]:
print('2018: ' + str(data_2018.shape))
print('2019: ' + str(data_2019.shape))
print('2020: ' + str(data_2020.shape))
print('2021: ' + str(data_2021.shape))

2018: (260432, 21)
2019: (124144, 21)
2020: (103208, 21)
2021: (23996, 21)


### **Save pre-processed data**

In [33]:
data_2018.to_csv("data_2018.csv", index = False)
data_2019.to_csv("data_2019.csv", index = False)
data_2020.to_csv("data_2020.csv", index = False)
data_2021.to_csv("data_2021.csv", index = False)

## **Change format of data from long to wide**

In [34]:
# Create df with all datasets
dfs = [data_2018, data_2019, data_2020, data_2021]
data = pd.concat(dfs)

In [35]:
# Create pivot table using the facility type
pivot_facility_type = pd.pivot_table(data, 
                       index=['section_2018','year','day_part'], 
                       columns=['facility_type'],
                       values=['cap_bike', 'cap_moped', 'cap_brokenup', 'cap_private', 
                               'bike', 'moped_yellow', 'moped_blue', 'cargo_bike', 'motor', 
                               'fourwheeler', 'rent_shared_bike','brokenup', 'stock', 
                               'capacity', 'occupancy'], 
                       aggfunc=np.sum,
                       fill_value=0)
pivot_facility_type.columns = pivot_facility_type.columns.to_flat_index().map(lambda x: '_in_'.join(x))
pivot_facility_type = pivot_facility_type.reset_index()
pivot_facility_type

section_2018  year day_part  bike_in_bike_shop  bike_in_mixed_space  \
0             10001  2018   middag                  0                    0   
1             10001  2019   middag                  0                    0   
2             10001  2021   middag                  0                    0   
3             10002  2018   middag                  0                    0   
4             10002  2019   middag                  0                    0   
...             ...   ...      ...                ...                  ...   
76548        500000  2020    avond                  0                    0   
76549        500000  2021   middag                  0                    0   
76550        500001  2019   middag                  0                    0   
76551        500001  2020    avond                  0                    0   
76552        500001  2021   middag                  0                    0   

       bike_in_moped_space  bike_in_no_facility  bike_in_other  bike_in_rack  \
0                        0                    0              0             0   
1                        0                    0              0             0   
2                        0                    0              0             0   
3                        0                    0              0             0   
4                        0                    0              0             0   
...                    ...                  ...            ...           ...   
76548                    0                    7              0             0   
76549                    0                    8              0             0   
76550                    0                    0              0             0   
76551                    0                    2              0             0   
76552                    0                    6              0             0   

       bike_in_space  bike_in_space_with_staples  bike_in_staples  \
0                  0                           0                0   
1                  0                           0                0   
2                  0                           0                0   
3                  0                           0                0   
4                  0                           0                0   
...              ...                         ...              ...   
76548              0                           0               19   
76549              0                           0               19   
76550              0                           0                0   
76551              0                           0                0   
76552              0                           0                0   

       brokenup_in_bike_shop  brokenup_in_mixed_space  \
0                          0                        0   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        0   
...                      ...                      ...   
76548                      0                        0   
76549                      0                        0   
76550                      0                        0   
76551                      0                        0   
76552                      0                        0   

       brokenup_in_moped_space  brokenup_in_no_facility  brokenup_in_other  \
0                            0                        0                  0   
1                            0                        0                  0   
2                            0                        0                  0   
3                            0                        0                  0   
4                            0                        0                  0   
...                        ...                      ...                ...   
76548                        0                   

In [36]:
# Create pivot table using the facility location
pivot_facility_location = pd.pivot_table(data, index=['section_2018','year','day_part'], 
                        columns=['facility_location'], 
                        values=['cap_bike', 'cap_moped', 'cap_brokenup', 'cap_private', 
                                'bike', 'moped_yellow', 'moped_blue', 'cargo_bike', 'motor', 
                                'fourwheeler', 'rent_shared_bike', 'brokenup','stock', 
                                'capacity', 'occupancy'],
                        aggfunc=np.sum, 
                        fill_value=0)
pivot_facility_location.columns = pivot_facility_location.columns.to_flat_index().map(lambda x: '_'.join(x))
pivot_facility_location = pivot_facility_location.reset_index()
pivot_facility_location

section_2018  year day_part  bike_inside  bike_outside  \
0             10001  2018   middag            0             0   
1             10001  2019   middag            0             0   
2             10001  2021   middag            0             0   
3             10002  2018   middag            0             0   
4             10002  2019   middag            0             0   
...             ...   ...      ...          ...           ...   
76548        500000  2020    avond           19             7   
76549        500000  2021   middag           19             8   
76550        500001  2019   middag            0             0   
76551        500001  2020    avond            0             2   
76552        500001  2021   middag            0             6   

       brokenup_inside  brokenup_outside  cap_bike_inside  cap_bike_outside  \
0                    0                 0                0                 0   
1                    0                 0                0                 0   
2                    0                 0                0                 0   
3                    0                 0                0                 0   
4                    0                 0                0                 0   
...                ...               ...              ...               ...   
76548                0                 0               20                 0   
76549                0                 0               20                 0   
76550                0                 0                0                 0   
76551                0                 0                0                 0   
76552                0                 0                0                 0   

       cap_brokenup_inside  cap_brokenup_outside  cap_moped_inside  \
0                        0                     0                 0   
1                        0                     0                 0   
2                        0                     0                 0   
3                        0                     0                 0   
4                        0                     0                 0   
...                    ...                   ...               ...   
76548                    0                     0                 0   
76549                    0                     0                 0   
76550                    0                     0                 0   
76551                    0                     0                 0   
76552                    0                     0                 0   

       cap_moped_outside  cap_private_inside  cap_private_outside  \
0                      0                   0                    0   
1                      0                   0                    0   
2                      0                   0                    0   
3                      0                   0                    0   
4                      0                   0                    0   
...                  ...                 ...                  ...   
76548                  0                   0                    0   
76549                  0                   0                    0   
76550                  0                   0                    0   
76551                  0                   0                    0   
76552                  0                   0                    0   

       capacity_inside  capacity_outside  cargo_bike_inside  \
0                    0                 0                  0   
1                    0                 0                  0   
2                    0                 0                  0   
3                    0                 0                  0   
4                    0                 0                  0   
...                ...               ...                ...   
76548               20                 0                  0   
76549               20                 0                  0   
76550                0                 0

In [37]:
# Create df with the sum of parking capacity and occupancy per location per time
grouped_per_location = data.groupby(['section_2018','year','day_part']).sum().reset_index()
grouped_per_location

section_2018  year day_part  cap_bike  cap_moped  cap_brokenup  \
0             10001  2018   middag         0          0             0   
1             10001  2019   middag         0          0             0   
2             10001  2021   middag         0          0             0   
3             10002  2018   middag         0          0             0   
4             10002  2019   middag         0          0             0   
...             ...   ...      ...       ...        ...           ...   
76548        500000  2020    avond        20          0             0   
76549        500000  2021   middag        20          0             0   
76550        500001  2019   middag         0          0             0   
76551        500001  2020    avond         0          0             0   
76552        500001  2021   middag         0          0             0   

       cap_private  bike  moped_yellow  moped_blue  cargo_bike  motor  \
0                0     0             0           0           0      0   
1                0     0             0           0           0      0   
2                0     0             0           0           0      0   
3                0     0             0           0           0      0   
4                0     0             0           0           0      0   
...            ...   ...           ...         ...         ...    ...   
76548            0    26             0           0           0      0   
76549            0    27             1           0           0      0   
76550            0     0             0           0           0      0   
76551            0     2             0           0           0      0   
76552            0     6             1           1           0      0   

       fourwheeler  rent_shared_bike  stock  brokenup  capacity  occupancy  
0                0                 0      0         0         0          0  
1                0                 0      0         0         0          0  
2                0                 0      0         0         0          0  
3                0                 0      0         0         0          0  
4                0                 0      0         0         0          0  
...            ...               ...    ...       ...       ...        ...  
76548            0                 0      0         0        20         26  
76549            0                 0      0         0        20         28  
76550            0                 0      0         0         0          0  
76551            0                 0      0         0         0          2  
76552            0                 0      0         0         0          8  

[76553 rows x 18 columns]

In [38]:
# Merge all three dataframes together
merged = pivot_facility_type.merge(pivot_facility_location, on=['section_2018','year','day_part']).merge(grouped_per_location, on=['section_2018','year','day_part'])
merged

section_2018  year day_part  bike_in_bike_shop  bike_in_mixed_space  \
0             10001  2018   middag                  0                    0   
1             10001  2019   middag                  0                    0   
2             10001  2021   middag                  0                    0   
3             10002  2018   middag                  0                    0   
4             10002  2019   middag                  0                    0   
...             ...   ...      ...                ...                  ...   
76548        500000  2020    avond                  0                    0   
76549        500000  2021   middag                  0                    0   
76550        500001  2019   middag                  0                    0   
76551        500001  2020    avond                  0                    0   
76552        500001  2021   middag                  0                    0   

       bike_in_moped_space  bike_in_no_facility  bike_in_other  bike_in_rack  \
0                        0                    0              0             0   
1                        0                    0              0             0   
2                        0                    0              0             0   
3                        0                    0              0             0   
4                        0                    0              0             0   
...                    ...                  ...            ...           ...   
76548                    0                    7              0             0   
76549                    0                    8              0             0   
76550                    0                    0              0             0   
76551                    0                    2              0             0   
76552                    0                    6              0             0   

       bike_in_space  bike_in_space_with_staples  bike_in_staples  \
0                  0                           0                0   
1                  0                           0                0   
2                  0                           0                0   
3                  0                           0                0   
4                  0                           0                0   
...              ...                         ...              ...   
76548              0                           0               19   
76549              0                           0               19   
76550              0                           0                0   
76551              0                           0                0   
76552              0                           0                0   

       brokenup_in_bike_shop  brokenup_in_mixed_space  \
0                          0                        0   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        0   
...                      ...                      ...   
76548                      0                        0   
76549                      0                        0   
76550                      0                        0   
76551                      0                        0   
76552                      0                        0   

       brokenup_in_moped_space  brokenup_in_no_facility  brokenup_in_other  \
0                            0                        0                  0   
1                            0                        0                  0   
2                            0                        0                  0   
3                            0                        0                  0   
4                            0                        0                  0   
...                        ...                      ...                ...   
76548                        0                   

### **Add more features to the merged dataset**

In [39]:
# Definition to add extra features
def add_features(df):
    df['perc_in'] = df['occupancy_inside'] / df['occupancy'] * 100
    df['perc_out'] = df['occupancy_outside'] / df['occupancy'] * 100
    df['net_point_occupancy'] = df['occupancy_inside'] / df['capacity'] * 100
    df['gross_point_occupancy'] = df['occupancy'] / df['capacity'] * 100
    
    df['perc_in'] = df['perc_in'].fillna(0)
    df['perc_out'] = df['perc_out'].fillna(0)
    df['net_point_occupancy'] = df['net_point_occupancy'].fillna(0)
    df['gross_point_occupancy'] = df['gross_point_occupancy'].fillna(0)

    df['gross_point_occupancy'].replace(np.inf, 250, inplace=True)
    df['net_point_occupancy'].replace(np.inf, 250, inplace=True)

    conditions = [
    (df['capacity'] == 0) & (df['occupancy_outside'] == 0),
    (df['capacity'] == 0) & (df['occupancy_outside'] > 0),
    (df['capacity'] > 0) & (df['occupancy_outside'] == 0),
    (df['capacity'] > 0) & (df['occupancy_inside'] < df['capacity']) & (df['occupancy_outside'] > 0),
    (df['capacity'] > 0) & (df['occupancy_inside'] == df['capacity']) & (df['occupancy_outside'] > 0),
    (df['capacity'] > 0) & (df['occupancy_inside'] > df['capacity']) & (df['occupancy_outside'] > 0),
    (df['capacity'] > 0) & (df['occupancy_inside'] < df['capacity']) & (df['occupancy_outside'] == 0),
    (df['capacity'] > 0) & (df['occupancy_inside'] == df['capacity']) & (df['occupancy_outside'] == 0)]
    choices_1 = ['No capacity, no occupancy','No capacity, but occupied','Capacity, but no occupancy, ','Capacity, but parked outside','Full capacity and parked outside','Overfull capacity and parked outside','Still capacity, no parked outside','Full capacity, no parked outside']
    choices_2 = [3,8,1,5,6,7,2,4]
    df['group_1'] = np.select(conditions, choices_1, default='?')
    df['group_2'] = np.select(conditions, choices_2, default='?')
    df['group_2'] = pd.to_numeric(df['group_2'])  

    return df

In [40]:
merged = add_features(merged)
merged

section_2018  year day_part  bike_in_bike_shop  bike_in_mixed_space  \
0             10001  2018   middag                  0                    0   
1             10001  2019   middag                  0                    0   
2             10001  2021   middag                  0                    0   
3             10002  2018   middag                  0                    0   
4             10002  2019   middag                  0                    0   
...             ...   ...      ...                ...                  ...   
76548        500000  2020    avond                  0                    0   
76549        500000  2021   middag                  0                    0   
76550        500001  2019   middag                  0                    0   
76551        500001  2020    avond                  0                    0   
76552        500001  2021   middag                  0                    0   

       bike_in_moped_space  bike_in_no_facility  bike_in_other  bike_in_rack  \
0                        0                    0              0             0   
1                        0                    0              0             0   
2                        0                    0              0             0   
3                        0                    0              0             0   
4                        0                    0              0             0   
...                    ...                  ...            ...           ...   
76548                    0                    7              0             0   
76549                    0                    8              0             0   
76550                    0                    0              0             0   
76551                    0                    2              0             0   
76552                    0                    6              0             0   

       bike_in_space  bike_in_space_with_staples  bike_in_staples  \
0                  0                           0                0   
1                  0                           0                0   
2                  0                           0                0   
3                  0                           0                0   
4                  0                           0                0   
...              ...                         ...              ...   
76548              0                           0               19   
76549              0                           0               19   
76550              0                           0                0   
76551              0                           0                0   
76552              0                           0                0   

       brokenup_in_bike_shop  brokenup_in_mixed_space  \
0                          0                        0   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        0   
...                      ...                      ...   
76548                      0                        0   
76549                      0                        0   
76550                      0                        0   
76551                      0                        0   
76552                      0                        0   

       brokenup_in_moped_space  brokenup_in_no_facility  brokenup_in_other  \
0                            0                        0                  0   
1                            0                        0                  0   
2                            0                        0                  0   
3                            0                        0                  0   
4                            0                        0                  0   
...                        ...                      ...                ...   
76548                        0                   

In [41]:
# Save the data
merged.to_csv("all_vehicle_parking_data.csv", index = False)